File: answers/go/src/problem_004.go
// Largest palindrome product

// A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

// Find the largest palindrome made from the product of two 3-digit numbers.

package main

import (
	"fmt"
	"strconv"
)

func main() {

	largestPalindrome := 0
	threeDigits := make([]int, 0)

	// build all possible threeDigits
	for i := 100; i < 1000; i++ {
		threeDigits = append(threeDigits, i)
	}

	for i := 0; i < len(threeDigits); i++ {
		for j := 0; j < len(threeDigits); j++ {
			product := threeDigits[i] * threeDigits[j]

			if isPalindromic(product) {
				if product > largestPalindrome {
					largestPalindrome = product
				}
			}
		}
	}

	fmt.Println(largestPalindrome) // 906609
}

func isPalindromic(x int) (result bool) {
	xString := strconv.Itoa(x)
	xLength := len(xString)
	xHalf := xLength/2 - 1

	xCheck := make([]int, 0)
	for i := 0; i < (xHalf + 1); i++ {
		xCheck = append(xCheck, i)
	}

	result = true

	for _, e := range xCheck {
		left := string(xString[e])
		right := string(xString[len(xString)-(1+e)])

		if left != right {
			result = false
		}
	}

	return
}

File: answers/scala/problem_004.scala
// Largest palindrome product

// A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

// Find the largest palindrome made from the product of two 3-digit numbers.

import scala.collection.mutable.ListBuffer

// functions

def intToList(number: Int): List[Int] = {
  val numberString = number.toString
  val length = numberString.length

  var list = ListBuffer[Int]()
  var i = 0
  do {
    val digit = Integer.parseInt( numberString.substring(i, i+1) )
    list += digit

    i += 1
  } while (i < length)

  list.toList
}

def isPalindrome(number: Int): Boolean = {
  val regularList = intToList(number)
  val reverseList = regularList.reverse

  regularList == reverseList
}

// find largest palindrome

var largestPalindrome = 0
val numbers = (100 to 999)

for (i <- numbers) {
  for (j <- numbers) {
    val product = i*j

    if (isPalindrome(product)) {
      if (product > largestPalindrome) {
        largestPalindrome = product
      }
    }
  }
}

println(largestPalindrome) //=> 906609
